Stats: Update paywall upgrade notice

a8ck3n opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to #92724 & pejTkB-1yy-p2.

## Proposed Changes

Updates the error state of the upgrade notice to explicitly call out when the paywall will go into effect if a commercial license is not purchased. Looks like this:

<img width="722" alt="SCR-20240729-phlx" src="https://github.com/user-attachments/assets/860e5001-c1e7-4859-a3c1-579715f4ff99">

Looking for feedback on the final copy before merging.

Current copy:

- title: You need to upgrade to a commercial license to continue using Jetpack Stats
- body: To ensure uninterrupted access to core Stats features, please upgrade to a Jetpack Stats Commercial license using the button below by LOCALIZED-DATE.

Note: The "by LOCALIZED-DATE" is omitted if the API doesn't provide it.

## Why are these changes being made?
<!--
It's easy to see what a PR does but much harder to find out why it was made,
particularly when researching old changes in history. Record an explanation of
the motivation behind this change and how it will help.
-->

If a site has been classified as commercial but does not yet have a commercial license, we grant a grace period before enabling the paywall. This notice is meant to give the user some heads up before they lose access to stats.

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Launch the Calypso Live branch.
* Visit a commercial site without an existing plan.
* Inspect the notice using the React dev tools. Should be an info notice if the blog sticker hasn't been added.
* Toggle the `showPaywallNotice` value in the inspector.
* Confirm you get the "error" version of the notice with more forceful messaging.

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

a8ck3n committed:
Force the paywall upgrade notice

a8ck3n committed:
Update title for error state

a8ck3n committed:
Update body for error state

The notice is now explicit about the features moving behind the paywall and surfaces the date in the message text.

a8ck3n committed:
Add emphasis to paywall date

a8ck3n committed:
Remove notice override

dognose24 commented on client/my-sites/stats/stats-notices/commercial-site-upgrade-notice.tsx:
```diff
@@ -24,6 +28,16 @@ const CommercialSiteUpgradeNotice = ( {
+	const moment = useLocalizedMoment();
+	const { data } = usePlanUsageQuery( siteId );
+	const dateFlagged = data?.paywall_date_from || moment().format( 'YYYY-MM-DD' );
+	const paywallFromDate = moment( dateFlagged )
```
The format of `paywall_date_from` is `YYYY-MM-DD HH:mm:ss`. If we add days to the date time and use the last date for the deadline, it would not be precise since we check the paywall with seconds.

For example, if the sticker was created at `2024-07-26 11:15:23`, the paywall will show after `2024-08-02 11:15:23`. And the deadline `2024-08-02` usually means `2024-08-02 23:59:59`.

So, I added a buffer day for the `upgrade_deadline_date` in the warning email trying to avoid users seeing the paywall before the end of the deadline date: D156486-code.

kangzj replied:
Makes sense to me! I guess we should localise the date, shouldn’t we? e.g.

`format('LLL')`

kangzj replied:
D156486-code adds a new property `upgrade_deadline_date` and maybe we should use that.

a8ck3n replied:
Thanks for the feedback. I'll get the date localized and change things over to use the `upgrade_deadline_date ` value. That should address the issue Dognose was mentioning as well.

kangzj commented on client/my-sites/stats/stats-notices/commercial-site-upgrade-notice.tsx:
```diff
@@ -76,8 +91,9 @@ const CommercialSiteUpgradeNotice = ( {
 	const bannerBody = showPaywallNotice
 		? translate(
-				'{{p}}Commercial sites with a significant number of visitors require a commercial license. Upgrade to get access to all the stats features and priority support.{{/p}}{{p}}{{jetpackStatsProductLink}}Upgrade my Stats{{/jetpackStatsProductLink}}{{commercialUpgradeLink}}{{commercialUpgradeLinkText}}Learn more{{/commercialUpgradeLinkText}}{{externalIcon /}}{{/commercialUpgradeLink}}{{/p}}',
+				'{{p}}Commercial sites require a commercial license. You will lose access to core Stats features on {{b}}%(date)s{{/b}}.{{/p}}{{p}}{{jetpackStatsProductLink}}Upgrade my Stats{{/jetpackStatsProductLink}}{{commercialUpgradeLink}}{{commercialUpgradeLinkText}}Learn more{{/commercialUpgradeLinkText}}{{externalIcon /}}{{/commercialUpgradeLink}}{{/p}}',
```
> Commercial sites require a commercial license. You will lose access to core Stats features on {{b}}%(date)s{{/b}}

I mean the action we are calling out to users is pretty obvious, but could we explicitly say that please? Something like:


> To ensure uninterrupted access to core Stats features, please upgrade to a Jetpack Stats Commercial license using the button below by {{b}}%(date)s{{/b}}. Thank you for your cooperation.

( ⬆️  by chatgpt)

a8ck3n replied:
Updated to use the new text but without the, "Thank you for your cooperation" bit.

a8ck3n committed:
Merge branch 'trunk' into update/stats-paywall-upgrade-notice

a8ck3n committed:
Localize displayed paywall date

a8ck3n committed:
Use upgrade_deadline_date value from API

No longer need to calculate an end date on the client side as the API provides this now.

a8ck3n committed:
Update comments, variable name

a8ck3n committed:
Guard against empty value from API

The value is defined as optional so we need to guard against that. Now we only show the date in the notice body if we have one.

dognose24 reviewed (approved):
As the date string `upgrade_deadline_date` has been localized within D156486-code. I used the string directly, and it worked well.

<img width="1275" alt="截圖 2024-07-29 下午11 06 09" src="https://github.com/user-attachments/assets/d3853fab-6d78-482b-9c29-d65260e6a88b">

<img width="1271" alt="截圖 2024-07-29 下午11 06 01" src="https://github.com/user-attachments/assets/6cb9149c-2a68-415e-ab15-14da35132272">

a8ck3n commented:
Looks good! Thanks for the fix, Dognose!

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/16434673).
